src/Microsoft.Extensions.OData.Migration/ODataMigrationMiddleware.cs

```diff
@@ -0,0 +1,170 @@
+﻿// ------------------------------------------------------------------------------
+// <copyright company="Microsoft Corporation">
+//     Copyright © Microsoft Corporation. All rights reserved.
+// </copyright>
+// ------------------------------------------------------------------------------
+
+namespace Microsoft.Extensions.OData.Migration
+{
+    using Microsoft.AspNetCore.Http;
+    using Microsoft.Data.OData.Query;
+    using Microsoft.Data.OData.Query.SemanticAst;
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.Specialized;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using System.Web;
+
+    /// <summary>
+    /// Translation Middleware currently converts V3 URI to V4 URI (future: converts query, request body as well)
+    /// </summary>
+    public class ODataMigrationMiddleware
+    {
+        private readonly RequestDelegate next;
+        private readonly Uri serviceRoot;
+        private readonly Data.Edm.IEdmModel v3Model;
+        private readonly Microsoft.OData.Edm.IEdmModel v4Model;
+
+        /// <summary>
+        /// Constructs an instance of TranslationMiddleware, requiring the root of the service, a V3 model instance and V4 model instance.
+        /// </summary>
+        /// <param name="next">Delegate required for middleware</param>
+        /// <param name="serviceRoot">Base path of service (e.g. "http://foobar:80/baz/")</param>
+        /// <param name="v3Model">Instance of V3 EDM model</param>
+        /// <param name="v4Model">Instance of V4 EDM model</param>
+        public ODataMigrationMiddleware(RequestDelegate next,
+                                         Uri serviceRoot,
+                                         Data.Edm.IEdmModel v3Model,
+                                         Microsoft.OData.Edm.IEdmModel v4Model)
+        {
+            this.next = next;
+            this.serviceRoot = serviceRoot;

```

Comment:
Is it needed to check if serviceRoot is null or empty?

Reply:
ODataUriParser will do the check, but I added my own for consistency, thanks!